core/src/qos_service.rs

```diff
@@ -177,24 +177,22 @@ impl QosService {
     pub fn update_or_remove_transaction_costs<'a>(
         transaction_costs: impl Iterator<Item = &'a TransactionCost>,
         transaction_qos_results: impl Iterator<Item = &'a transaction::Result<()>>,
-        retryable_transaction_indexes: &[usize],
+        execution_results: impl Iterator<Item = &'a TransactionExecutionResult>,
         bank: &Arc<Bank>,
     ) {
         let mut cost_tracker = bank.write_cost_tracker().unwrap();
         transaction_costs
             .zip(transaction_qos_results)
-            .enumerate()
-            .for_each(|(index, (tx_cost, qos_inclusion_result))| {
-                // Only transactions that the qos service incuded have been added to the
-                // cost tracker.
-                if qos_inclusion_result.is_ok() && retryable_transaction_indexes.contains(&index) {
-                    cost_tracker.remove(tx_cost);
-                } else {
-                    // TODO: Update the cost tracker with the actual execution compute units.
-                    // Will have to plumb it in next; For now, keep estimated costs.
-                    //
-                    // let actual_execution_cost = 0;
-                    // cost_tracker.update_execution_cost(tx_cost, actual_execution_cost);
+            .zip(execution_results)
+            .for_each(|((tx_cost, qos_inclusion_result), execution_result)| {
+                // Only transactions that the qos service included have to be
+                // checked for remove or update/commit
+                if qos_inclusion_result.is_ok() {
+                    if execution_result.was_executed_successfully() {
```

Comment:
It sounds to me like we should use `was_executed()` here: failed transactions should contribute to block cost.

Reply:
AFAICS I think the PR and the issue mention removing `non-retryable failed` from costs, this code is removing`non-retryable failed` & `retryable`, it should only remove `retryable` which is !was_executed() as ckamm mentions

Reply:
good call, fixed.